# Análise dos dados das variáveis relacionadas ao IDEB

O notebook contém experimentos com essas variáveis. São utilizadas duas bases de dados:<br>
- Váriáveis derivadas do SAEB (informações socioeconômicas)<br> 
- Variáveis com informações das escolas e professores<br>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
from sklearn.feature_selection import SelectKBest, RFE, RFECV, SelectFromModel
from sklearn.feature_selection import mutual_info_classif, f_classif, chi2
from sklearn.preprocessing import LabelEncoder
import numpy as np
from numpy import array 
from sklearn.naive_bayes import MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import sklearn.metrics as metrics

## Importando bases de dados

In [ ]:
#Importando variáveis derivadas do SAEB

#base_estudantes_se = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/variaveis_derivadas_se.csv')

#base_estudantes_se = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_se.csv')

#base_estudantes_se.set_index('ID_ESCOLA', inplace=True)

#base_estudantes_al = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/variaveis_derivadas_al.csv')

#base_estudantes_al = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_al.csv')

#base_estudantes_al.set_index('ID_ESCOLA', inplace=True)

base_estudantes_ba = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_ba.csv')

dados_pessoais = base_estudantes_ba

#base_estudantes_ba.set_index('ID_ESCOLA', inplace=True)

#dados_pessoais = pd.merge(base_estudantes_se, base_estudantes_al, how = 'outer')

#dados_pessoais = pd.merge(base_estudantes_ba, dados_pessoais, how = 'outer')

In [ ]:
dados_pessoais.set_index('ID_ESCOLA', inplace=True)

In [ ]:
dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
29211590              2017          2  ...               6               7
29032539              2017          2  ...              14              13
29033047              2017          2  ...              10               6
29446740              2017          2  ...               7               4
61206519              2017          2  ...               7               5
...                    ...        ...  ...             ...             ...
61218410              2017          2  ...              12              12
61227355              2017          2  ...               1               1
61238810              2017          2  ...               2               0
61213619              2017          2  ...               2               1
61215745              2017          2  ...              13              12

[1144 rows x 235 columns]

**Pré-processamento da base de dados pessoais**

In [ ]:
#Função que verifica se o ID de uma escola é inválido
#Segundo o Dicionário de Dados da base o SAEB 6 - Os códigos dos Municípios e 
#das Escolas que começam com o dígito "6" são máscaras, isto é, são códigos fictícios.
def retira_id_ficticio(id_escola):
    if id_escola[0] == "6":
        return True
    else:
        return False

In [ ]:
#Se retirou as escolas fictícias
for escola in dados_pessoais.index:
    if (retira_id_ficticio(str(escola))):
        dados_pessoais.drop(escola, inplace=True)
        

dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
29211590              2017          2  ...               6               7
29032539              2017          2  ...              14              13
29033047              2017          2  ...              10               6
29446740              2017          2  ...               7               4
29236606              2017          2  ...              14               6
...                    ...        ...  ...             ...             ...
29409500              2017          2  ...               8               6
29433410              2017          2  ...               8              16
29456487              2017          2  ...               7               3
29457491              2017          2  ...               8               5
29041333              2017          2  ...               4               8

[396 rows x 235 columns]

In [ ]:
#Importando variáveis com informações das escolas e professores
#base_escola_se = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-SE.csv')

##base_escola_se = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-SE.csv')

#base_escola_se.set_index('CO_ENTIDADE', inplace=True)

#base_escola_al = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-AL.csv')

#base_escola_al = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-AL.csv')

#base_escola_al.set_index('CO_ENTIDADE', inplace=True)

#base_escola_ba = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-BA.csv')

base_escola_ba = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-BA.csv')

base_escola_ba.rename(columns={'Unnamed: 0': 'CO_ENTIDADE'}, inplace=True)

base_escola_ba.set_index('CO_ENTIDADE', inplace=True)

#dados_escola = pd.merge(base_escola_se, base_escola_al, how = 'outer')

#dados_escola = pd.concat([dados_escola, base_escola_ba], join='outer')

dados_escola = base_escola_ba

In [ ]:
#dados_escola.set_index('CO_ENTIDADE', inplace=True)

dados_escola

NU_ANO_CENSO                                        NO_ENTIDADE  \
CO_ENTIDADE                                                                    
29433410             2017  EE - COLEGIO ESTADUAL DOM CLIMERIO ALMEIDA DE ...   
29208149             2017  EE - COLEGIO ESTADUAL ARISTIDES SILVA - DISTRI...   
29526604             2017      EE - COLEGIO ESTADUAL NOSSA SENHORA APARECIDA   
29424348             2017            EE - COLEGIO ESTADUAL ANTONIO FIGUEREDO   
29378052             2017                 CENTRO EDUCACIONAL MONTEIRO LOBATO   
...                   ...                                                ...   
29223687             2017  EE - COLEGIO ESTADUAL EDIVALDO BOAVENTURA - PO...   
29035007             2017  EE - COLEGIO ESTADUAL POLIVALENTE DE PAULO AFONSO   
29242290             2017         EE - INSTITUTO DE EDUCACAO ANISIO TEIXEIRA   
29237645             2017              EE - COLEGIO ESTADUAL JOAO VILAS BOAS   
29337941             2017       EE - COLEGIO ESTADUAL ISABEL ARAUJO DA SILVA   

             CO_ORGAO_REGIONAL  TP_SITUACAO_FUNCIONAMENTO  \
CO_ENTIDADE                                                 
29433410                  20.0                          1   
29208149                   1.0                          1   
29526604                  11.0                          1   
29424348                  13.0                          1   
29378052                  13.0                          1   
...                        ...                        ...   
29223687                   9.0                          1   
29035007                  24.0                          1   
29242290                  13.0                          1   
29237645                  13.0                          1   
29337941                  11.0                          1   

            DT_ANO_LETIVO_INICIO DT_ANO_LETIVO_TERMINO  CO_REGIAO  \
CO_ENTIDADE                                                         
29433410              06/02/2017            22/12/2017          2   
29208149              06/02/2017            13/12/2017          2   
29526604              06/02/2017            13/12/2017          2   
29424348              06/02/2017            13/12/2017          2   
29378052              06/02/2017            30/11/2017          2   
...                          ...                   ...        ...   
29223687              06/02/2017            13/12/2017          2   
29035007              06/02/2017            13/12/2017          2   
29242290              06/02/2017            13/12/2017          2   
29237645              06/02/2017            13/12/2017          2   
29337941              06/02/2017            13/12/2017          2   

             CO_MESORREGIAO  CO_MICRORREGIAO  CO_UF  ...  QtdeDocentes  \
CO_ENTIDADE                                          ...                 
29433410               2906            29028     29  ...           214   
29208149               2906            29022     29  ...            39   
29526604               2901            29001     29  ...           164   
29424348               2906            29026     29  ...           171   
29378052               2906            29027     29  ...           176   
...                     ...              ...    ...  ...           ...   
29223687               2906            29024     29  ...           184   
29035007               2902            29005     29  ...           332   
29242290               2906            29026     29  ...           483   
29237645               2906            29025     29  ...           361   
29337941               2901            29001     29  ...           186   

             QtdeDocentesAuxAssist  QtdeDocentesMonit  asQtdeDocentesEfetivos  \
CO_ENTIDADE                                                                     
29433410                         0                  0                     214   
29208149                         0                  0                     

In [ ]:
col_ideb = dados_escola['Ideb-2017']

dados_escola.drop('Ideb-2017', axis=1, inplace=True)

In [ ]:
col_ideb

CO_ENTIDADE
29433410    3.3
29208149    4.1
29526604    2.9
29424348    3.6
29378052    6.4
           ... 
29223687    2.9
29035007    3.9
29242290    4.0
29237645    3.7
29337941    2.5
Name: Ideb-2017, Length: 431, dtype: float64

In [ ]:
dados_escola

NU_ANO_CENSO  ... QtdeAlunosTotal
CO_ENTIDADE                ...                
29433410             2017  ...             822
29208149             2017  ...             214
29526604             2017  ...             727
29424348             2017  ...             707
29378052             2017  ...             911
...                   ...  ...             ...
29223687             2017  ...             797
29035007             2017  ...            1512
29242290             2017  ...            2009
29237645             2017  ...            1580
29337941             2017  ...             620

[431 rows x 183 columns]

#### Pré-processamento da base das escolas

In [ ]:
#Retirando valores ausentes
filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

# retorna o número de valores ausentes para cada coluna
dados_escola[listOfColumnNames].isnull().sum()

CO_ORGAO_REGIONAL    7
dtype: int64

In [ ]:
#Retirando colunas que apresentam mais de 60% do valores nulos
dados_escola.dropna(thresh=len(dados_escola)*0.6, axis=1, inplace=True)

#Preenchendo valores ausentes
dados_escola.fillna(0, inplace=True)

filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

# retorna o número de valores ausentes para cada coluna
dados_escola[listOfColumnNames].isnull().sum()

CO_ORGAO_REGIONAL    0
dtype: int64

In [ ]:
dados_escola[listOfColumnNames] = dados_escola[listOfColumnNames].astype(int)

In [ ]:
dados_escola.insert(len(dados_escola.columns), 'Ideb-2017', col_ideb)
#['Ideb-2017'] = col_ideb

In [ ]:
dados_escola

NU_ANO_CENSO  ... Ideb-2017
CO_ENTIDADE                ...          
29433410             2017  ...       3.3
29208149             2017  ...       4.1
29526604             2017  ...       2.9
29424348             2017  ...       3.6
29378052             2017  ...       6.4
...                   ...  ...       ...
29223687             2017  ...       2.9
29035007             2017  ...       3.9
29242290             2017  ...       4.0
29237645             2017  ...       3.7
29337941             2017  ...       2.5

[431 rows x 184 columns]

In [ ]:
#Colunas categóricas
filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.object]

listOfColumnNames = list(filteredColumns.index)

dados_escola[listOfColumnNames]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  


NO_ENTIDADE  ... DT_ANO_LETIVO_TERMINO
CO_ENTIDADE                                                     ...                      
29433410     EE - COLEGIO ESTADUAL DOM CLIMERIO ALMEIDA DE ...  ...            22/12/2017
29208149     EE - COLEGIO ESTADUAL ARISTIDES SILVA - DISTRI...  ...            13/12/2017
29526604         EE - COLEGIO ESTADUAL NOSSA SENHORA APARECIDA  ...            13/12/2017
29424348               EE - COLEGIO ESTADUAL ANTONIO FIGUEREDO  ...            13/12/2017
29378052                    CENTRO EDUCACIONAL MONTEIRO LOBATO  ...            30/11/2017
...                                                        ...  ...                   ...
29223687     EE - COLEGIO ESTADUAL EDIVALDO BOAVENTURA - PO...  ...            13/12/2017
29035007     EE - COLEGIO ESTADUAL POLIVALENTE DE PAULO AFONSO  ...            13/12/2017
29242290            EE - INSTITUTO DE EDUCACAO ANISIO TEIXEIRA  ...            13/12/2017
29237645                 EE - COLEGIO ESTADUAL JOAO VILAS BOAS  ...            13/12/2017
29337941          EE - COLEGIO ESTADUAL ISABEL ARAUJO DA SILVA  ...            13/12/2017

[431 rows x 3 columns]

In [ ]:
#Tratando datas, convertendo para timestamp

inicio = pd.to_datetime(dados_escola['DT_ANO_LETIVO_INICIO'], infer_datetime_format = True)
termino = pd.to_datetime(dados_escola['DT_ANO_LETIVO_TERMINO'], infer_datetime_format = True)

dados_escola['DT_ANO_LETIVO_INICIO'] = inicio.values.astype (np.int64)
dados_escola['DT_ANO_LETIVO_TERMINO'] = termino.values.astype (np.int64)

In [ ]:
dados_escola.drop('NO_ENTIDADE', axis=1, inplace=True)

In [ ]:
dados_escola.columns.to_list()

['NU_ANO_CENSO',
 'CO_ORGAO_REGIONAL',
 'TP_SITUACAO_FUNCIONAMENTO',
 'DT_ANO_LETIVO_INICIO',
 'DT_ANO_LETIVO_TERMINO',
 'CO_REGIAO',
 'CO_MESORREGIAO',
 'CO_MICRORREGIAO',
 'CO_UF',
 'CO_MUNICIPIO',
 'CO_DISTRITO',
 'TP_DEPENDENCIA',
 'TP_LOCALIZACAO',
 'TP_CATEGORIA_ESCOLA_PRIVADA',
 'IN_CONVENIADA_PP',
 'TP_CONVENIO_PODER_PUBLICO',
 'IN_MANT_ESCOLA_PRIVADA_EMP',
 'IN_MANT_ESCOLA_PRIVADA_ONG',
 'IN_MANT_ESCOLA_PRIVADA_SIND',
 'IN_MANT_ESCOLA_PRIVADA_SIST_S',
 'IN_MANT_ESCOLA_PRIVADA_S_FINS',
 'CO_ESCOLA_SEDE_VINCULADA',
 'CO_IES_OFERTANTE',
 'TP_REGULAMENTACAO',
 'IN_LOCAL_FUNC_PREDIO_ESCOLAR',
 'TP_OCUPACAO_PREDIO_ESCOLAR',
 'IN_LOCAL_FUNC_SALAS_EMPRESA',
 'IN_LOCAL_FUNC_SOCIOEDUCATIVO',
 'IN_LOCAL_FUNC_UNID_PRISIONAL',
 'IN_LOCAL_FUNC_PRISIONAL_SOCIO',
 'IN_LOCAL_FUNC_TEMPLO_IGREJA',
 'IN_LOCAL_FUNC_CASA_PROFESSOR',
 'IN_LOCAL_FUNC_GALPAO',
 'TP_OCUPACAO_GALPAO',
 'IN_LOCAL_FUNC_SALAS_OUTRA_ESC',
 'IN_LOCAL_FUNC_OUTROS',
 'IN_PREDIO_COMPARTILHADO',
 'IN_AGUA_FILTRADA',
 'IN_AGUA_RE

In [ ]:
dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
29211590              2017          2  ...               6               7
29032539              2017          2  ...              14              13
29033047              2017          2  ...              10               6
29446740              2017          2  ...               7               4
29236606              2017          2  ...              14               6
...                    ...        ...  ...             ...             ...
29409500              2017          2  ...               8               6
29433410              2017          2  ...               8              16
29456487              2017          2  ...               7               3
29457491              2017          2  ...               8               5
29041333              2017          2  ...               4               8

[396 rows x 235 columns]

### Unindo as bases de dados 

Foi realizado join nos dois datasets, nesse momento se percebeu que não existem variáveis escolas particulares na base derivada do SAEB

In [ ]:
dataset = dados_pessoais.join(dados_escola, how="inner")

In [ ]:
dataset

ID_PROVA_BRASIL  ID_REGIAO  ...  QtdeAlunosTotal  Ideb-2017
29211590             2017          2  ...              287        3.7
29032539             2017          2  ...              685        3.5
29033047             2017          2  ...              442        3.1
29446740             2017          2  ...              689        2.5
29236606             2017          2  ...              551        3.1
...                   ...        ...  ...              ...        ...
29409500             2017          2  ...              877        2.9
29433410             2017          2  ...              822        3.3
29456487             2017          2  ...              370        3.2
29457491             2017          2  ...              268        3.1
29041333             2017          2  ...              975        1.9

[393 rows x 418 columns]

In [ ]:
dados_escola.dtypes[dados_escola.dtypes == np.float64]

Ideb-2017    float64
dtype: object

In [ ]:
dataset['ID_DEPENDENCIA_ADM'].unique()

array([2, 3, 1])

In [ ]:
#dataset.to_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa Ideb/variaveis_derivadas_escola_estudante_5.0.csv')
dataset.to_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_escola_estudante_5.0.csv')

### Discretização dos rótulos, nota do Ideb

**Discretização:** É o processo de transferência de funções contínuas, modelos, variáveis e equações em contrapartes discretas.

In [ ]:
# São três classes média, acima da média e abaixo da média
def discretizacao(ideb, mean_ideb):
    classe = []
    for nota in ideb['Ideb-2017']:
        if (mean_ideb == nota):
            classe.append('média')
        elif (mean_ideb < nota):
            classe.append('acima da média')
        else:
            classe.append('abaixo da média')
    return classe

In [ ]:
#discretizando as notas do Ideb

Ideb = dataset[['Ideb-2017']]

mean_Ideb = round(Ideb.mean(),1)

classe_ideb = discretizacao(Ideb, mean_Ideb['Ideb-2017'])

Ideb.insert(1, 'classe-Ideb', classe_ideb)

Ideb

Ideb-2017      classe-Ideb
29211590        3.7   acima da média
29032539        3.5   acima da média
29033047        3.1            média
29446740        2.5  abaixo da média
29236606        3.1            média
...             ...              ...
29409500        2.9  abaixo da média
29433410        3.3   acima da média
29456487        3.2   acima da média
29457491        3.1            média
29041333        1.9  abaixo da média

[393 rows x 2 columns]

In [ ]:
#Variavel a ser prevista
y = Ideb['classe-Ideb']

**Seleção de Atributos:**  É um passo da fase de pré-processamento que tem a finalidade de selecionar o melhor subconjunto de atributos que pode substituir o conjunto de dados original. 

https://www.analyticsvidhya.com/blog/2016/12/introduction-to-feature-selection-methods-with-an-example-or-how-to-select-the-right-variables/#:~:text=Embedded%20methods%20combine%20the%20qualities,penalization%20functions%20to%20reduce%20overfitting.

Abordagens:

- Filtro: a relevância dos atributos é avaliada independentemente – não
envolve o uso de um método de AM ;

- Wrapper (Envólucro): Avalia o subconjunto de atributos arti culado ao método de AM que realizará a indução do conceito;

- Embedded (Embutidos): Os métodos incorporados combinam as qualidades dos métodos de filtro e wrapper. Ele é implementado por algoritmos que possuem seus próprios métodos de seleção de recursos integrados.

In [ ]:
#Filter based
# f_classif: ANOVA F-value between label/feature for classification tasks.
#ANOVA stands for Analysis of variance. It is similar to LDA except for the fact that it is operated using one or more categorical 
#independent features and one continuous dependent feature. It provides a statistical test 
#of whether the means of several groups are equal or not.
def selecao_atributos(X, y, quant):
    select = SelectKBest(score_func=f_classif, k=quant)
    #select = SelectKBest(score_func=chi2, k=quant)
    z = select.fit_transform(X, y)
    
    filter = select.get_support(indices=True)
    features = array(X.columns)
    
    return X.iloc[:, filter]

In [ ]:
#Wrapper-based
#from sklearn.cluster import KMeans

def wrapper_based(X, X_transform, y, quant):

  #select_features = RFE(estimator=KMeans(n_clusters=3), n_features_to_select=100)
  select_features = RFE(estimator=SVC(kernel="linear"), n_features_to_select=100)
  '''select_features = RFECV(
      estimator=SVC(kernel="linear"),
      min_features_to_select= quant,
      cv = 5
  )'''
  select_features.fit(X_transform, y)

  filter = select_features.get_support(indices=True)
  return X.iloc[:, filter]

In [ ]:
# Embedded-based

def embedded_based(X, X_transform, y, quant):

  select_features = SelectFromModel(LogisticRegression(), max_features=quant)
  #select_features = SelectFromModel(DecisionTreeClassifier(), max_features=quant)
  #select_features = SelectFromModel(SVC(kernel="rbf"), max_features=quant)
  #select_features = SelectFromModel(RandomForestClassifier(), max_features=quant)
  select_features.fit(X_transform, y)

  filter = select_features.get_support(indices=True)

  return X.iloc[:, filter]

**Métricas de Avaliação**


In [ ]:
def acuracia(resultados, classes):
    #resultados = cross_val_predict(modelo, textos, classes, cv=10)  
    return 'Acurácia do modelo: {} %'.format(round(metrics.accuracy_score(classes, resultados)*100, 2))

**Normalização:** é mudar os valores das colunas numéricas no conjunto de dados para usar uma escala comum, sem distorcer as diferenças nos intervalos de valores nem perder informações.

- **Normalização MinMax:** Para cada recurso, o valor mínimo desse recurso é transformado em 0, o valor máximo é transformado em 1 e todos os outros valores são transformados em um decimal entre 0 e 1.

In [ ]:
def minmax(data):
    scaler = MinMaxScaler()
    data_scaler = scaler.fit_transform(data)
    return data_scaler

## Criação dos modelos

**Serão feitas três análises:** Apenas com as variáveis socioeconômicas, com as variáveis de escolas e professores e depois com o dataset completo

Para cada uma serão criados modelos com todos os atributos e também com seleção de atributos, a fim de verificar o de maior acurácia

In [ ]:
def inicializar_modelos():
    naive = MultinomialNB()
    random = RandomForestClassifier(n_estimators=150)
    decision_tree = DecisionTreeClassifier()
    svm = SVC(kernel="linear")
    r_logistic = LogisticRegression(max_iter=250)
    knn = KNeighborsClassifier(n_neighbors=3)
    return naive, random, decision_tree, svm, r_logistic, knn

#### -Primeira análise: variáveis socioeconômicas derivadas da prova SAEB**

In [ ]:
#Foi retirada colunas do 'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF'
columns = dataset.columns[3:234]

columns

Index(['ID_MUNICIPIO', 'ID_AREA', 'ID_DEPENDENCIA_ADM', 'ID_LOCALIZACAO',
       'IN_SITUACAO_CENSO', 'TX_RESP_Q001_A', 'TX_RESP_Q001_B',
       'TX_RESP_Q020_A', 'TX_RESP_Q020_B', 'TX_RESP_Q021_A',
       ...
       'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D',
       'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H',
       'TX_RESP_Q003_I', 'TX_RESP_Q003_J'],
      dtype='object', length=231)

In [ ]:
X =  dataset[columns]

In [ ]:
#Questões de sexo, raça, ano, mes de nascimento, conformidade com censo
question = ['TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D', 
            'TX_RESP_Q004_E', 'TX_RESP_Q004_F', 'TX_RESP_Q004_G', 'TX_RESP_Q004_H', 'IN_SITUACAO_CENSO']

delete_at = ['IN_SITUACAO_CENSO']

X.drop(delete_at, axis=1, inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
filteredColumns = X.dtypes[X.dtypes == np.float]

listOfColumnNames = list(filteredColumns.index)

X[listOfColumnNames]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: []
Index: [29211590, 29032539, 29033047, 29446740, 29236606, 29425417, 29150450, 29395712, 29417163, 29456800, 29459010, 29292280, 29222583, 29157722, 29395950, 29434785, 29047501, 29426618, 29005752, 29089964, 29090350, 29151554, 29223652, 29255597, 29461464, 29417880, 29465400, 29886732, 29161770, 29464765, 29080967, 29036143, 29213592, 29385253, 29395860, 29000831, 29001498, 29383218, 29059380, 29294037, 29268524, 29081319, 29081440, 29081815, 29450837, 29043085, 29385334, 29461499, 29214211, 29214254, 29203643, 29384400, 29384389, 29223687, 29427541, 29205000, 29294606, 29440173, 29163226, 29240638, 29242290, 29433568, 29069106, 29351502, 29447348, 29048605, 29050057, 29461510, 29176808, 29383234, 29415969, 29243882, 29437164, 29398282, 29070317, 29409535, 29044570, 29019508, 29163293, 29163374, 29158419, 29158850, 29033187, 29424526, 29468000, 29445949, 29090865, 29164826, 29137900, 29138434, 29138442, 29139155, 29432200, 29258324, 29214866, 29092205, 29259630, 29012465, 29013259, 29013275, ...]

[393 rows x 0 columns]

**Divisão entre conjunto de treino e teste**

In [ ]:
X_scaler = minmax(X)

X_train, X_test, y_train, y_test = train_test_split(X_scaler, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 67.09 %
Naive Bayes: Acurácia do modelo: 63.29 %
Decision Tree: Acurácia do modelo: 58.23 %
SVM: Acurácia do modelo: 64.56 %
Logistic Regression: Acurácia do modelo: 64.56 %
KNN: Acurácia do modelo: 48.1 %


**-Primeira Análise: Aplicando Seleção de Atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X, y, 100)

X_selection

ID_AREA  ID_DEPENDENCIA_ADM  ...  TX_RESP_Q004_G  TX_RESP_Q004_H
29211590        2                   2  ...               1               8
29032539        2                   2  ...               7              12
29033047        2                   2  ...               9               7
29446740        2                   2  ...               3               3
29236606        2                   2  ...               5               2
...           ...                 ...  ...             ...             ...
29409500        2                   2  ...               3               7
29433410        2                   2  ...               4               6
29456487        2                   2  ...               7               8
29457491        2                   2  ...               6              14
29041333        2                   2  ...               4              13

[393 rows x 100 columns]

In [ ]:
X_scaler = minmax(X_selection)

X_train, X_test, y_train, y_test = train_test_split(X_scaler, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 64.56 %
Naive Bayes: Acurácia do modelo: 64.56 %
Decision Tree: Acurácia do modelo: 49.37 %
SVM: Acurácia do modelo: 63.29 %
Logistic Regression: Acurácia do modelo: 65.82 %
KNN: Acurácia do modelo: 55.7 %


**-Primeira Análise: Aplicando Seleção de Atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_scaler, y, 100)

X_embedded

ID_MUNICIPIO  ID_AREA  ...  TX_RESP_Q039_B  TX_RESP_Q039_C
29211590       2900108        2  ...              11               1
29032539       2900207        2  ...              25               4
29033047       2900207        2  ...              22               2
29446740       2900306        2  ...              10               3
29236606       2900504        2  ...              11               2
...                ...      ...  ...             ...             ...
29409500       2933307        2  ...               8               1
29433410       2933307        2  ...              17               2
29456487       2933307        2  ...              64               1
29457491       2933307        2  ...              35               3
29041333       2933604        2  ...              19               3

[393 rows x 42 columns]

In [ ]:
X_scaler_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_scaler_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 59.49 %
Naive Bayes: Acurácia do modelo: 64.56 %
Decision Tree: Acurácia do modelo: 59.49 %
SVM: Acurácia do modelo: 70.89 %
Logistic Regression: Acurácia do modelo: 70.89 %
KNN: Acurácia do modelo: 58.23 %


**-Primeira Análise: Aplicando Seleção de Atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_scaler, y, 100)

X_wrapper

ID_MUNICIPIO  ID_AREA  ...  TX_RESP_Q045_C  TX_RESP_Q045_D
29211590       2900108        2  ...              20               1
29032539       2900207        2  ...              38               3
29033047       2900207        2  ...              33               1
29446740       2900306        2  ...              31               1
29236606       2900504        2  ...              43               2
...                ...      ...  ...             ...             ...
29409500       2933307        2  ...              19               0
29433410       2933307        2  ...              44               4
29456487       2933307        2  ...              23               0
29457491       2933307        2  ...              11               2
29041333       2933604        2  ...              25               3

[393 rows x 100 columns]

In [ ]:
X_scaler_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_scaler_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 63.29 %
Naive Bayes: Acurácia do modelo: 65.82 %
Decision Tree: Acurácia do modelo: 51.9 %
SVM: Acurácia do modelo: 70.89 %
Logistic Regression: Acurácia do modelo: 69.62 %
KNN: Acurácia do modelo: 62.03 %


#### -Segunda análise: Análise com o dataset completo

In [ ]:
#Lembrar de tratar datas

'''question = ['TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q003_L', 'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D', 'CO_REGIAO',
            'TP_SITUACAO_FUNCIONAMENTO', 'TP_REGULAMENTACAO',
            'CO_MESORREGIAO', 'CO_MICRORREGIAO', 'CO_DISTRITO', 'TP_LOCALIZACAO',  'CO_ORGAO_REGIONAL', 'TP_DEPENDENCIA',
            'TX_RESP_Q004_E', 'TX_RESP_Q004_F', 'TX_RESP_Q004_G', 'TX_RESP_Q004_H', 'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF', 'ID_MUNICIPIO', 
            'DT_ANO_LETIVO_INICIO', 'DT_ANO_LETIVO_TERMINO', 'IN_SITUACAO_CENSO', 'NU_ANO_CENSO', 'CO_UF', 'CO_MUNICIPIO', 'Ideb-2017']'''

'''delete = ['CO_MESORREGIAO', 'CO_MICRORREGIAO', 'CO_DISTRITO', 'TP_LOCALIZACAO',  'CO_ORGAO_REGIONAL', 'TP_DEPENDENCIA',
          'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF', 'ID_MUNICIPIO', 'IN_SITUACAO_CENSO', 'Ideb-2017']'''


#X = dataset.drop(delete, axis=1)

'''X = dataset.drop(['Ideb-2017','IN_SITUACAO_CENSO', 'TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q003_L', 'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D',], axis=1)'''

X = dataset.drop(['Ideb-2017','IN_SITUACAO_CENSO'], axis=1)

In [ ]:
list(X.columns)

['ID_PROVA_BRASIL',
 'ID_REGIAO',
 'ID_UF',
 'ID_MUNICIPIO',
 'ID_AREA',
 'ID_DEPENDENCIA_ADM',
 'ID_LOCALIZACAO',
 'TX_RESP_Q001_A',
 'TX_RESP_Q001_B',
 'TX_RESP_Q020_A',
 'TX_RESP_Q020_B',
 'TX_RESP_Q021_A',
 'TX_RESP_Q021_B',
 'TX_RESP_Q024_A',
 'TX_RESP_Q024_B',
 'TX_RESP_Q025_A',
 'TX_RESP_Q025_B',
 'TX_RESP_Q027_A',
 'TX_RESP_Q027_B',
 'TX_RESP_Q028_A',
 'TX_RESP_Q028_B',
 'TX_RESP_Q029_A',
 'TX_RESP_Q029_B',
 'TX_RESP_Q030_A',
 'TX_RESP_Q030_B',
 'TX_RESP_Q031_A',
 'TX_RESP_Q031_B',
 'TX_RESP_Q038_A',
 'TX_RESP_Q038_B',
 'TX_RESP_Q043_A',
 'TX_RESP_Q043_B',
 'TX_RESP_Q044_A',
 'TX_RESP_Q044_B',
 'TX_RESP_Q052_A',
 'TX_RESP_Q052_B',
 'TX_RESP_Q018_A',
 'TX_RESP_Q018_B',
 'TX_RESP_Q018_C',
 'TX_RESP_Q022_A',
 'TX_RESP_Q022_B',
 'TX_RESP_Q022_C',
 'TX_RESP_Q026_A',
 'TX_RESP_Q026_B',
 'TX_RESP_Q026_C',
 'TX_RESP_Q032_A',
 'TX_RESP_Q032_B',
 'TX_RESP_Q032_C',
 'TX_RESP_Q033_A',
 'TX_RESP_Q033_B',
 'TX_RESP_Q033_C',
 'TX_RESP_Q034_A',
 'TX_RESP_Q034_B',
 'TX_RESP_Q034_C',
 'TX_RESP_Q

In [ ]:
X

ID_PROVA_BRASIL  ...  QtdeAlunosTotal
29211590             2017  ...              287
29032539             2017  ...              685
29033047             2017  ...              442
29446740             2017  ...              689
29236606             2017  ...              551
...                   ...  ...              ...
29409500             2017  ...              877
29433410             2017  ...              822
29456487             2017  ...              370
29457491             2017  ...              268
29041333             2017  ...              975

[393 rows x 416 columns]

In [ ]:
filteredColumns = X.dtypes[X.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

#X[listOfColumnNames].fillna(-1, inplace=True)

#X[listOfColumnNames] = X[listOfColumnNames].astype(int)

X[listOfColumnNames].isnull().sum()

Series([], dtype: float64)

In [ ]:
X_transform = minmax(X)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 68.35 %
Naive Bayes: Acurácia do modelo: 64.56 %
Decision Tree: Acurácia do modelo: 59.49 %
SVM: Acurácia do modelo: 63.29 %
Logistic Regression: Acurácia do modelo: 67.09 %
KNN: Acurácia do modelo: 59.49 %


**-Segunda Análise: Aplicando Seleção de Atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_transform, y, 100)

X_embedded

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


ID_DEPENDENCIA_ADM  ...  QtdeAlunosTotal
29211590                   2  ...              287
29032539                   2  ...              685
29033047                   2  ...              442
29446740                   2  ...              689
29236606                   2  ...              551
...                      ...  ...              ...
29409500                   2  ...              877
29433410                   2  ...              822
29456487                   2  ...              370
29457491                   2  ...              268
29041333                   2  ...              975

[393 rows x 100 columns]

In [ ]:
X_transform_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_transform_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 68.35 %
Naive Bayes: Acurácia do modelo: 64.56 %
Decision Tree: Acurácia do modelo: 51.9 %
SVM: Acurácia do modelo: 68.35 %
Logistic Regression: Acurácia do modelo: 68.35 %
KNN: Acurácia do modelo: 56.96 %


**-Segunda Análise: Aplicando Seleção de Atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_transform, y, 100)

X_wrapper

ID_DEPENDENCIA_ADM  ...  asQtdeDocentesEfetivos
29211590                   2  ...                     110
29032539                   2  ...                      95
29033047                   2  ...                      69
29446740                   2  ...                      69
29236606                   2  ...                      57
...                      ...  ...                     ...
29409500                   2  ...                     167
29433410                   2  ...                     214
29456487                   2  ...                      99
29457491                   2  ...                      60
29041333                   2  ...                     196

[393 rows x 100 columns]

In [ ]:
X_transform_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_transform_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 65.82 %
Naive Bayes: Acurácia do modelo: 68.35 %
Decision Tree: Acurácia do modelo: 54.43 %
SVM: Acurácia do modelo: 69.62 %
Logistic Regression: Acurácia do modelo: 70.89 %
KNN: Acurácia do modelo: 60.76 %


**-Segunda Análise: Aplicando Seleção de Atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X, y, 100)

X_selection

/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [  0   1   2 234 236 239 242 247 248 249 250 251 252 253 254 258 260 261
 262 263 264 265 278 280 303 355 363 364 365 369 370 372 373 382 383 384
 385 386 387 388 389 393 394 395 397 403 404] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


ID_DEPENDENCIA_ADM  ...  asQtdeDocentesEfetivos
29211590                   2  ...                     110
29032539                   2  ...                      95
29033047                   2  ...                      69
29446740                   2  ...                      69
29236606                   2  ...                      57
...                      ...  ...                     ...
29409500                   2  ...                     167
29433410                   2  ...                     214
29456487                   2  ...                      99
29457491                   2  ...                      60
29041333                   2  ...                     196

[393 rows x 100 columns]

In [ ]:
X_transform = minmax(X_selection)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 93.38 %
Naive Bayes: Acurácia do modelo: 63.36 %
Decision Tree: Acurácia do modelo: 91.35 %
SVM: Acurácia do modelo: 74.3 %
Logistic Regression: Acurácia do modelo: 74.55 %
KNN: Acurácia do modelo: 70.74 %


**Segunda Análise: Unindo subconjunto de melhores atributos**

In [ ]:
X_all = pd.concat([X_selection, X_wrapper], axis=1)

X_all = pd.concat([X_all, X_embedded], axis=1)

In [ ]:
#Removendo colunas duplicadas
X_new_all = X_all.T.drop_duplicates().T

In [ ]:
X_transform = minmax(X_new_all)
#X_transform = minmax(X_all)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 93.38 %
Naive Bayes: Acurácia do modelo: 64.89 %
Decision Tree: Acurácia do modelo: 92.37 %
SVM: Acurácia do modelo: 82.44 %
Logistic Regression: Acurácia do modelo: 82.19 %
KNN: Acurácia do modelo: 71.76 %


Resultado com Todos atributos

- Random forest: Acurácia do modelo: 93.64 %
- Naive Bayes: Acurácia do modelo: 64.89 %
- Decision Tree: Acurácia do modelo: 91.35 %
- SVM: Acurácia do modelo: 83.97 %
- Logistic Regression: Acurácia do modelo: 83.46 %
- KNN: Acurácia do modelo: 74.05 %

**Segunda Análise: Aplicando a técnica merge**

Link da Dissertação: https://www.dcc.ufmg.br/pos/cursos/defesas/1930M.PDF


In [ ]:
#Metodo Merge combina os atributos mais frequentes nos melhores conjuntos de seleção 
#Cada atributo irá apresentar um score de acordo com sua presença em um subconjunto de dados
def metodo_merge(selection, embedded, wrapper, ranking):
  for atributo in selection:
    score = ranking[atributo]
    if (atributo in embedded):
      score += 1
      embedded.remove(atributo)
    if (atributo in wrapper):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score

  for atributo in embedded:
    score = ranking[atributo]
    if (atributo in selection):
      score += 1
      embedded.remove(atributo)
    if (atributo in wrapper):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score

  for atributo in wrapper:
    score = ranking[atributo]
    if (atributo in selection):
      score += 1
      embedded.remove(atributo)
    if (atributo in embedded):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score
 
  return ranking      

In [ ]:
#Gerando Listas com os melhores atributos de cada abordagem
selection = list(X_selection.columns)
embedded = list(X_embedded.columns)
wrapper = list(X_wrapper.columns)

#Criando dicionário
values = [1]*300

all = selection + embedded + wrapper

ranking = dict(zip(all, values))

ranking = metodo_merge(selection, embedded, wrapper, ranking)

#Filtrando os atributos que estão em mais de um subconjunto

dic_aux = ranking.items()

filtered = {key:value for key, value in dic_aux if value > 1}

#Gerando X após o merge

X_merge = dataset[list(filtered.keys())]

In [ ]:
X_transform = minmax(X_merge)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 93.89 %
Naive Bayes: Acurácia do modelo: 67.43 %
Decision Tree: Acurácia do modelo: 90.84 %
SVM: Acurácia do modelo: 79.9 %
Logistic Regression: Acurácia do modelo: 78.12 %
KNN: Acurácia do modelo: 72.77 %


In [ ]:
dados_para_exportar = pd.DataFrame(columns=X_merge.columns, data=X_merge.values, index=X_merge.index)
dados_para_exportar.insert(len(dados_para_exportar.columns), "Valor Real Ideb", y.to_numpy())
dados_para_exportar.insert(len(dados_para_exportar.columns), "Previsão Ideb", random_result)


dados_para_exportar

ID_DEPENDENCIA_ADM  TX_RESP_Q020_B  ...  Valor Real Ideb    Previsão Ideb
29211590                   2               7  ...   acima da média   acima da média
29032539                   2              23  ...   acima da média  abaixo da média
29033047                   2              21  ...            média  abaixo da média
29446740                   2              11  ...  abaixo da média  abaixo da média
29236606                   2              17  ...            média            média
...                      ...             ...  ...              ...              ...
29409500                   2              10  ...  abaixo da média  abaixo da média
29433410                   2               8  ...   acima da média   acima da média
29456487                   2              18  ...   acima da média  abaixo da média
29457491                   2              16  ...            média            média
29041333                   2              27  ...  abaixo da média  abaixo da média

[393 rows x 88 columns]

In [ ]:
dados_para_exportar = dados_para_exportar.loc[dados_para_exportar['Valor Real Ideb'] ==  dados_para_exportar['Previsão Ideb']]


dados_para_exportar

ID_DEPENDENCIA_ADM  TX_RESP_Q020_B  ...  Valor Real Ideb    Previsão Ideb
29211590                   2               7  ...   acima da média   acima da média
29446740                   2              11  ...  abaixo da média  abaixo da média
29236606                   2              17  ...            média            média
29425417                   2               1  ...            média            média
29150450                   2               0  ...   acima da média   acima da média
...                      ...             ...  ...              ...              ...
29361540                   1               0  ...   acima da média   acima da média
29409500                   2              10  ...  abaixo da média  abaixo da média
29433410                   2               8  ...   acima da média   acima da média
29457491                   2              16  ...            média            média
29041333                   2              27  ...  abaixo da média  abaixo da média

[369 rows x 88 columns]

In [ ]:
#dados_para_exportar.to_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/resultado_analise_merge_dataset_completo.csv')
#dados_para_exportar.to_csv('/content/drive/MyDrive/TCC/resultado_analise_merge_dataset_completo.csv')
dados_para_exportar.to_csv('/content/drive/MyDrive/TCC/resultado_analise_merge_dataset_completo_BA.csv')

#### -Terceira análise: variáveis das escolas e professores

In [ ]:
columns_escolas = dataset.columns[237:404]

X_escolas = dataset[columns_escolas]

X_escolas

TP_SITUACAO_FUNCIONAMENTO  ...  QtdeDocentesFCEM
29211590                          1  ...                 3
29032539                          1  ...                 3
29033047                          1  ...                 5
29446740                          1  ...                 0
29236606                          1  ...                 0
...                             ...  ...               ...
29409500                          1  ...                12
29433410                          1  ...                 7
29456487                          1  ...                 4
29457491                          1  ...                 0
29041333                          1  ...                 0

[393 rows x 167 columns]

In [ ]:
X_escolas.drop(columns=['DT_ANO_LETIVO_INICIO', 'DT_ANO_LETIVO_TERMINO'], axis=1, inplace=True)

X_transform = minmax(X_escolas)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 64.56 %
Naive Bayes: Acurácia do modelo: 60.76 %
Decision Tree: Acurácia do modelo: 54.43 %
SVM: Acurácia do modelo: 58.23 %
Logistic Regression: Acurácia do modelo: 63.29 %
KNN: Acurácia do modelo: 58.23 %


**Aplicando seleção de atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X_escolas, y, 100)

/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [  0   1   4   9  10  11  12  13  14  15  16  20  22  23  24  25  26  27
  40  42  65 117 125 126 127 131 132 134 135 144 145 146 147 148 149 150
 151 155 156 157 159] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 67.09 %
Naive Bayes: Acurácia do modelo: 60.76 %
Decision Tree: Acurácia do modelo: 49.37 %
SVM: Acurácia do modelo: 58.23 %
Logistic Regression: Acurácia do modelo: 63.29 %
KNN: Acurácia do modelo: 58.23 %


**Aplicando seleção de atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_transform, y, 100)

X_embedded

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


ID_UF  ID_DEPENDENCIA_ADM  ...  TX_RESP_Q013_C  TX_RESP_Q013_E
29211590     29                   2  ...               2               0
29032539     29                   2  ...               7               0
29033047     29                   2  ...               0               0
29446740     29                   2  ...               6               1
29236606     29                   2  ...               1               0
...         ...                 ...  ...             ...             ...
29409500     29                   2  ...               2               0
29433410     29                   2  ...              11               0
29456487     29                   2  ...               2               0
29457491     29                   2  ...               1               0
29041333     29                   2  ...               0               0

[393 rows x 78 columns]

In [ ]:
X_transform_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_transform_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 63.29 %
Naive Bayes: Acurácia do modelo: 63.29 %
Decision Tree: Acurácia do modelo: 65.82 %
SVM: Acurácia do modelo: 69.62 %
Logistic Regression: Acurácia do modelo: 64.56 %
KNN: Acurácia do modelo: 55.7 %


**Aplicando seleção de atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_transform, y, 100)

X_wrapper

ID_UF  ID_MUNICIPIO  ...  TX_RESP_Q013_D  TX_RESP_Q014_A
29211590     29       2900108  ...               1               0
29032539     29       2900207  ...               1               6
29033047     29       2900207  ...               0               3
29446740     29       2900306  ...               0               1
29236606     29       2900504  ...               0               1
...         ...           ...  ...             ...             ...
29409500     29       2933307  ...               1               0
29433410     29       2933307  ...               2               0
29456487     29       2933307  ...               0               0
29457491     29       2933307  ...               1               2
29041333     29       2933604  ...               0              12

[393 rows x 100 columns]

In [ ]:
X_transform_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_transform_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 64.56 %
Naive Bayes: Acurácia do modelo: 64.56 %
Decision Tree: Acurácia do modelo: 55.7 %
SVM: Acurácia do modelo: 69.62 %
Logistic Regression: Acurácia do modelo: 70.89 %
KNN: Acurácia do modelo: 56.96 %
